src/VisualQueryBuilder/components/LabelFilterItem.tsx

```diff
@@ -0,0 +1,216 @@
+import  { uniqBy } from 'lodash';
+import React, { useState } from 'react';
+
+import { SelectableValue, toOption } from '@grafana/data';
+import { selectors } from '@grafana/e2e-selectors';
+import { InlineField, Select } from '@grafana/ui';
+
+import { QueryBuilderLabelFilter } from '../types';
+import { InputGroup } from '../../QueryEditor/InputGroup';
+import { AccessoryButton } from '../../QueryEditor/AccessoryButton';
+
+interface Props {
+  defaultOp: string;
+  item: Partial<QueryBuilderLabelFilter>;
+  items: Array<Partial<QueryBuilderLabelFilter>>;
+  onChange: (value: QueryBuilderLabelFilter) => void;
+  onGetLabelNames: (forLabel: Partial<QueryBuilderLabelFilter>) => Promise<SelectableValue[]>;
+  onGetLabelValues: (forLabel: Partial<QueryBuilderLabelFilter>) => Promise<SelectableValue[]>;
+  onDelete: () => void;
+  invalidLabel?: boolean;
+  invalidValue?: boolean;
+  multiValueSeparator?: string;
+}
+
+export function LabelFilterItem({
+  item,
+  items,
+  defaultOp,
+  onChange,
+  onDelete,
+  onGetLabelNames,
+  onGetLabelValues,
+  invalidLabel,
+  invalidValue,
+  multiValueSeparator = "|",
+}: Props) {
+  const [state, setState] = useState<{
+    labelNames?: SelectableValue[];
+    labelValues?: SelectableValue[];
+    isLoadingLabelNames?: boolean;
+    isLoadingLabelValues?: boolean;
+  }>({});
+  // there's a bug in react-select where the menu doesn't recalculate its position when the options are loaded asynchronously
+  // see https://github.com/grafana/grafana/issues/63558
+  // instead, we explicitly control the menu visibility and prevent showing it until the options have fully loaded
+  const [labelNamesMenuOpen, setLabelNamesMenuOpen] = useState(false);
+  const [labelValuesMenuOpen, setLabelValuesMenuOpen] = useState(false);
```

Comment:
Another thing I would fix if we follow this up is to either use a single object `useState` or multiple individual `useState`, not a mix.